File: map-gen-script/src/map_generator/topologies/square.py
# src/map_generator/topologies/square.py

import random
from .base_topology import BaseTopology
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_X, FORWARD_Z, BACKWARD_X, BACKWARD_Z

class SquareTopology(BaseTopology):
    """
    Tạo ra một con đường hình vuông.
    
    Đây là dạng map kinh điển để dạy về việc lặp lại một chuỗi hành động
    (ví dụ: đi thẳng một đoạn rồi rẽ). Nó sử dụng logic "vòng lặp mở",
    nghĩa là điểm bắt đầu và kết thúc nằm bên ngoài hình vuông.
    """

    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể hình vuông với kích thước cạnh khác nhau.
        """
        count = 0
        # Lặp qua các kích thước cạnh có thể có
        for side_length in range(3, 9):
            if count >= max_variants: return
            
            variant_params = copy.deepcopy(params)
            variant_params['side_length'] = side_length
            yield self.generate_path_info(grid_size, variant_params)
            count += 1
    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """
        Tạo ra một hình vuông với kích thước và vị trí ngẫu nhiên.

        Args:
            params (dict): Có thể chứa 'side_length' để tùy chỉnh kích thước.

        Returns:
            PathInfo: Một đối tượng chứa thông tin về hình vuông. 
                      'path_coords' sẽ chứa tọa độ của 4 góc.
        """
        print("    LOG: Generating 'square_shape' topology...")
        
        # Lấy kích thước cạnh, mặc định là 5. side_length là số bước đi.
        # (CẢI TIẾN) Đọc và ngẫu nhiên hóa kích thước cạnh từ params
        side_length_param = params.get('side_length', [4, 6]) # Mặc định là một khoảng
        if isinstance(side_length_param, list) and len(side_length_param) == 2:
            # Nếu side_length là một khoảng [min, max], chọn ngẫu nhiên một giá trị
            side_length = random.randint(side_length_param[0], side_length_param[1])
        else:
            # Nếu không, sử dụng giá trị được cung cấp
            side_length = int(side_length_param)

        # Đảm bảo hình vuông nằm gọn trong map với một lề an toàn
        # Cần side_length + 1 ô cho các góc, và thêm 1 ô cho start/target
        max_size = min(grid_size[0], grid_size[2]) - 2
        if side_length > max_size:
            side_length = max_size
        if side_length < 2: side_length = 2 # Đảm bảo kích thước tối thiểu
            
        start_x = random.randint(1, grid_size[0] - side_length - 2)
        start_z = random.randint(1, grid_size[2] - side_length - 2)
        y = 0

        # --- Xác định vị trí bắt đầu và kết thúc theo logic "vòng lặp mở" ---
        # Người chơi bắt đầu ở một ô bên trái, đi vào góc C1.
        start_pos: Coord = (start_x - 1, y, start_z)
        target_pos: Coord = (start_x - 1, y, start_z + side_length)

        # --- (SỬA LỖI) Tạo ra path_coords và placement_coords đầy đủ ---
        path_coords: list[Coord] = []
        
        # Điểm bắt đầu và kết thúc cũng là một phần của đường đi
        path_coords.append(start_pos)

        # Tạo đường đi đầy đủ cho 4 cạnh
        current_pos = (start_x, y, start_z)
        path_coords.append(current_pos)
        c1 = current_pos
        for _ in range(side_length):
            current_pos = add_vectors(current_pos, FORWARD_X)
            path_coords.append(current_pos)
        c2 = current_pos
        for _ in range(side_length):
            current_pos = add_vectors(current_pos, FORWARD_Z)
            path_coords.append(current_pos)
        c3 = current_pos
        for _ in range(side_length):
            current_pos = add_vectors(current_pos, BACKWARD_X)
            path_coords.append(current_pos)
        c4 = current_pos
        
        # Thêm đường đi ra target
        path_coords.append(target_pos)

        # placement_coords chỉ chứa 4 góc để Placer đặt vật phẩm
        placement_coords = list(dict.fromkeys(path_coords)) # Sử dụng toàn bộ đường đi cho placement_coords để render ground

        # [NEW] Tạo metadata cho Placers
        corners = [c1, c2, c3, c4]
        
        # Tạo segments từ các góc
        segments = []
        # Segment 1: c1 -> c2
        seg1 = [(c1[0] + i, y, c1[2]) for i in range(side_length + 1)]
        segments.append(seg1)
        # Segment 2: c2 -> c3
        seg2 = [(c2[0], y, c2[2] + i) for i in range(side_length + 1)]
        segments.append(seg2)
        # Segment 3: c3 -> c4
        seg3 = [(c3[0] - i, y, c3[2]) for i in range(side_length + 1)]
        segments.append(seg3)
        # Segment 4: c4 -> c1 (optional, closes the square)
        seg4 = [(c4[0], y, c4[2] - i) for i in range(side_length + 1)]
        segments.append(seg4)
        
        metadata = {
            "segments": segments,
            "corners": corners,
            "side_length": side_length,
        }

        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=list(dict.fromkeys(path_coords)), # Loại bỏ các điểm trùng lặp
            placement_coords=placement_coords,
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/Square.ts
/**
 * Square Topology (PORTED FROM PYTHON)
 * Creates a square path using 'Open Loop' logic.
 * Path Start/End are outside the square loop.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';
import { addVectors, FORWARD_X, FORWARD_Z, BACKWARD_X } from '../utils/geometry';

export class SquareTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;
    for (let side = 3; side < 9; side++) {
        if (count >= maxVariants) return;
        const variantParams = { ...params };
        variantParams.side_length = side;
        yield this.generatePathInfo(gridSize, variantParams);
        count++;
    }
  }

  generatePathInfo(
    gridSize: [number, number, number],
    params: Record<string, any>
  ): IPathInfo {
    console.log("    LOG: Generating 'square_shape' topology...");

    let sideLength = 4; // Default
    const sideParam = params.side_length || [4, 6];
    if (Array.isArray(sideParam)) {
        sideLength = Math.floor(Math.random() * (sideParam[1] - sideParam[0] + 1)) + sideParam[0];
    } else {
        sideLength = Number(sideParam);
    }

    const maxSize = Math.min(gridSize[0], gridSize[2]) - 2;
    if (sideLength > maxSize) sideLength = maxSize;
    if (sideLength < 2) sideLength = 2;

    const startX = Math.floor(Math.random() * (gridSize[0] - sideLength - 2)) + 1;
    const startZ = Math.floor(Math.random() * (gridSize[2] - sideLength - 2)) + 1;
    const y = 0;

    // Open Loop Logic
    // Start is outside (left of C1)
    // C1 is (startX, y, startZ)
    const startPos: Coord = [startX - 1, y, startZ];
    const targetPos: Coord = [startX - 1, y, startZ + sideLength];

    const pathCoords: Coord[] = [];
    
    // 1. Start Point
    pathCoords.push(startPos);

    // 2. Square Loop 4 Sides (Partial)
    let currentPos: Coord = [startX, y, startZ];
    pathCoords.push([...currentPos]); // C1
    const c1 = [...currentPos];

    // Side 1 (X+)
    for (let i = 0; i < sideLength; i++) {
        currentPos = addVectors(currentPos, FORWARD_X);
        pathCoords.push(currentPos);
    }
    const c2 = [...currentPos];

    // Side 2 (Z+)
    for (let i = 0; i < sideLength; i++) {
        currentPos = addVectors(currentPos, FORWARD_Z);
        pathCoords.push(currentPos);
    }
    const c3 = [...currentPos];

    // Side 3 (X-)
    for (let i = 0; i < sideLength; i++) {
        currentPos = addVectors(currentPos, BACKWARD_X);
        pathCoords.push(currentPos);
    }
    const c4 = [...currentPos];

    // Side 4 (Z-) - Not fully traversed in Python!
    // Python Square code: Segments 1, 2, 3... and seg 4.
    // Wait, Python code has:
    // for _ in range(side_length): FORWARD_X
    // for _ in range(side_length): FORWARD_Z
    // for _ in range(side_length): BACKWARD_X
    // Then append target_pos.
    // It DOES NOT do the 4th side (Z-)!
    // It creates a U-shape square (3 sides) and then jumps to target?
    // Target pos: (start_x - 1, y, start_z + side_length) -> This is near C4?
    // C4 is (startX, y, startZ + side.length).
    // Target is (startX - 1, --). So it's just to the left of C4.
    // So distinct segments are 3 sides.
    
    pathCoords.push(targetPos);

    // Python Metadata includes seg4 (c4->c1) as optional closed square
    // seg4 = [(c4[0], y, c4[2] - i) ...]
    // But path_coords doesn't include it. 
    // Wait, Python code Line 114: `seg4` is created in metadata list but NOT in `path_coords` loop.
    
    // Placement Coords: Python uses `list(dict.fromkeys(path_coords))`
    // So placement is just the path (U shape).

    // --- RECONSTRUCT METADATA ---
    const corners = [c1, c2, c3, c4];
    
    // Segments for metadata (Python constructs them mathematically in metadata block)
    const segments: Coord[][] = [];
    
    // Seg 1: C1 -> C2
    const seg1: Coord[] = [];
    for(let i=0; i<=sideLength; i++) seg1.push([c1[0]+i, y, c1[2]]);
    segments.push(seg1);

    // Seg 2: C2 -> C3
    const seg2: Coord[] = [];
    for(let i=0; i<=sideLength; i++) seg2.push([c2[0], y, c2[2]+i]);
    segments.push(seg2);

    // Seg 3: C3 -> C4
    const seg3: Coord[] = [];
    for(let i=0; i<=sideLength; i++) seg3.push([c3[0]-i, y, c3[2]]);
    segments.push(seg3);

    // Seg 4: C4 -> C1 (Ghost segment for metadata)
    const seg4: Coord[] = [];
    for(let i=0; i<=sideLength; i++) seg4.push([c4[0], y, c4[2]-i]);
    segments.push(seg4);

    const lengths = segments.map(s => s.length);

    const metadata = {
        topology_type: "square_shape", // Python says 'square_shape'
        segments: segments,
        corners: corners,
        side_length: sideLength,
        segment_analysis: {
            count: segments.length,
            lengths: lengths,
            min_length: Math.min(...lengths),
            max_length: Math.max(...lengths),
            min_valid_range: Math.max(0, Math.min(...lengths) - 2),
            total_valid_slots: lengths.reduce((sum, l) => sum + Math.max(0, l - 2), 0),
        },
        semantic_positions: {
            start: startPos,
            end: targetPos,
            c1: c1,
            c2: c2,
            c3: c3,
            c4: c4,
            optimal_start: 'start',
            optimal_end: 'end',
            valid_pairs: [
                {
                    name: 'full_square_easy',
                    start: 'start',
                    end: 'end',
                    path_type: 'full_traversal',
                    strategies: ['segment_pattern_reuse', 'corner_based'],
                    difficulty: 'EASY',
                    teaching_goal: 'Identical items per side'
                },
                {
                    name: 'corner_emphasis_medium',
                    start: 'c1',
                    end: 'c3',
                    path_type: 'diagonal_corners',
                    strategies: ['segment_pattern_reuse', 'corner_logic'],
                    difficulty: 'MEDIUM',
                    teaching_goal: 'Special corner items'
                },
                {
                    name: 'progressive_sides_hard',
                    start: 'start',
                    end: 'end',
                    path_type: 'progressive_sides',
                    strategies: ['segment_pattern_reuse', 'side_identical'],
                    difficulty: 'HARD',
                    teaching_goal: 'Progressive pattern per side'
                }
            ]
        }
    };

    return {
        start_pos: startPos,
        target_pos: targetPos,
        path_coords: Array.from(new Set(pathCoords.map(c => c.join(',')))).map(s => s.split(',').map(Number) as Coord),
        placement_coords: Array.from(new Set(pathCoords.map(c => c.join(',')))).map(s => s.split(',').map(Number) as Coord),
        obstacles: [],
        metadata: metadata
    };
  }
}


